src/main/java/org/snomed/snowstorm/core/data/services/traceability/TraceabilityConsumer.java

```diff
@@ -1,23 +1,56 @@
 package org.snomed.snowstorm.core.data.services.traceability;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.context.annotation.Lazy;
 import org.springframework.jms.core.JmsTemplate;
 import org.springframework.stereotype.Component;
 
+import java.util.List;
+
 @Component
 @Lazy
 public class TraceabilityConsumer {
 
 	@Value("${jms.queue.prefix}")
 	private String jmsQueuePrefix;
 
+	@Value("${activemq.max.message.concept-activities}")
+	private int maxConceptActiviesPerMessage;
+
 	@Autowired
 	private JmsTemplate jmsTemplate;
 
+	private final Logger logger = LoggerFactory.getLogger(getClass());
+
 	public void accept(Activity activity) {
-		jmsTemplate.convertAndSend(jmsQueuePrefix + ".traceability", activity);
+		if (activity.getChanges().size() <= maxConceptActiviesPerMessage) {
+			jmsTemplate.convertAndSend(jmsQueuePrefix + ".traceability", activity);
+		}
+		else {
+			sendInBatches(activity);
+		}
 	}
 
-}
+	/**
+	 * A large activity object (with too many changes) should be messaged in batches.
+	 * @param activity
+	 */
+	void sendInBatches(Activity activity) {
+		int changeListSize = activity.getChanges().size();
+		logger.info("Number of concept activities is {}. larger than max ({}).", changeListSize, maxConceptActiviesPerMessage );
+		int idx = 0;
+
+		while (idx <= activity.getChanges().size() -1 ) {
```

Comment:
@markrynbeek  The logic is sound however we could make it simpler by using com.google.common.collect.Iterables.partition() method. The alternative batching is based on changes for N concepts rather than N changes.

Reply:
I'll use that partition method, thanks. As for N: each change list element is a concept plus its component changes, so N changes and N concepts amounts to the same.